### DISCUSSION TITLE
How to validate data depending on another property

### DISCUSSION THREAD

**Discussion by ssbarnea** [pydantic/pydantic#2498] [Question] [answered]
I have a list of platforms and with a list of valid platforms for platform, as seen at https://github.com/ansible-community/schemas/blob/main/src/ansibleschemas/_galaxy.py

I do want to enable validation of these values, so users will be notified if they put an invalid version for a specific platform.

Current model looks like:
```python
class PlatformModel(BaseModel):
    name: str
    versions: List[str]  # user can specify a list of versions
```

I do know how to make platform be an enum, but this would not allow me to limit the versions to allow only those versions permitted by that specific platform.

**Comment by PrettyWood** [comment, marked as answer]
Hi @ssbarnea 
You could just use `validator`
```py
from pydantic import BaseModel, validator

VALID_VERSIONS = {'q', 'w'}

class PlatformModel(BaseModel):
    name: str
    versions: list[str]

    @validator('versions', each_item=True)
    def valid_version(cls, v: str) -> str:
        if v not in VALID_VERSIONS:
            raise ValueError(f'{v!r} is not a valid version')
        return v


PlatformModel(name='qwe', versions=['a', 'w', 'b', 'q'])
# pydantic.error_wrappers.ValidationError: 2 validation errors for PlatformModel
# versions -> 0
#   'a' is not a valid version (type=value_error)
# versions -> 2
#   'b' is not a valid version (type=value_error)
```
Hope it helps!

**Reply by ssbarnea** [reply]
Is this going to endup in the generated schema or not? If not I will go for generating models for each platform, using create_model. My use-case is to produce a schema, not to use Python to perform the validation, at least at this point.

**Reply by PrettyWood** [reply]
What about
```py
from typing import Literal

from pydantic import BaseModel

VALID_VERSIONS = {'q', 'w'}

class PlatformModel(BaseModel):
    name: str
    versions: Literal[tuple(VALID_VERSIONS)]
```
?
If this doesn't help you please provide a scenario with expected behaviour

**Reply by ssbarnea** [reply]
I ended up dynamically building a type for each platform where each platform contains its own enum inside. The resulting schema validates files correctly.

I am no proud about the [current code](https://github.com/ansible-community/schemas/blob/main/src/ansibleschemas/meta.py#L52) but I am mentioning here in case someone else needs some help.

When you start to dynamically create types, mypy starts screaming and I was forced to add some type ignore after I failed to find a way to make it happy.
